src/sinks/datadog/traces/request_builder.rs

```diff
@@ -0,0 +1,351 @@
+                                })),
+                            }
+                        }
+                        Err(err) => results.push(Err(RequestBuilderError::FailedToEncode {
+                            reason: err.parts().0,
+                            dropped_events: err.parts().1,
+                        })),
+                    })
+            }
+        }
+        results
+    }
+
+    fn build_request(&mut self, metadata: Self::Metadata, payload: Self::Payload) -> Self::Request {
+        let mut headers = BTreeMap::<String, String>::new();
+        headers.insert(
+            "Content-Type".to_string(),
+            "application/x-protobuf".to_string(),
+        );
+        headers.insert("DD-API-KEY".to_string(), metadata.api_key.to_string());
+        headers.insert(
+            "X-Datadog-Reported-Languages".to_string(),
+            metadata.lang.unwrap_or_else(|| "".into()),
+        );
+        if let Some(ce) = self.compression.content_encoding() {
+            headers.insert("Content-Encoding".to_string(), ce.to_string());
+        }
+        TraceApiRequest {
+            batch_size: metadata.batch_size,
+            body: payload,
+            headers,
+            finalizers: metadata.finalizers,
+            uri: self
+                .endpoint_configuration
+                .get_uri_for_endpoint(metadata.endpoint),
+        }
+    }
+}
+
+pub struct DatadogTracesEncoder {
+    max_size: usize,
+}
+
+#[derive(Debug, Snafu)]
+pub enum EncoderError {
+    #[snafu(display("Unable to split payload into small enough chunks"))]
+    UnableToSplit { dropped_events: u64 },
+}
+
+impl EncoderError {
+    pub const fn parts(&self) -> (&'static str, u64) {
+        match self {
+            Self::UnableToSplit { dropped_events: n } => ("unable to split into small chunks", *n),
+        }
+    }
+}
+impl DatadogTracesEncoder {
+    fn encode_trace(
+        &self,
+        key: &PartitionKey,
+        events: Vec<TraceEvent>,
+    ) -> Vec<Result<(Vec<u8>, Vec<TraceEvent>), EncoderError>> {
+        let mut encoded_payloads = Vec::new();
+        let payload = DatadogTracesEncoder::trace_into_payload(key, &events);
+        let encoded_payload = payload.encode_to_vec();
+        // This may happen exceptionally
+        if encoded_payload.len() > self.max_size {
+            emit!(DatadogTracesPayloadTooBig {});
+            let n_chunks: usize = (encoded_payload.len() / self.max_size) + 1;
+            let chunk_size = (events.len() / n_chunks) + 1;
+            events.chunks(chunk_size).for_each(|events| {
+                let chunked_payload = DatadogTracesEncoder::trace_into_payload(key, events);
+                let encoded_chunk = chunked_payload.encode_to_vec();
+                if encoded_chunk.len() > self.max_size {
+                    encoded_payloads.push(Err(EncoderError::UnableToSplit {
+                        dropped_events: events.len() as u64,
+                    }));
```

Comment:
When would this happen in practice? If a single trace is too big?

Reply:
Yes exactly.

Reply:
Could we make the internal event error message more explicit then? We could have it log that it received a trace that is greater than the configured batch size and so dropped it. It could also have an `error_code` of something like `message_too_big`.